### ISSUE TITLE
Duplicate field + getter definitions share the same getter element

### ISSUE THREAD

**Issue by DanTup** [dart-lang/sdk#49547]
Given this code (from `DuplicateDefinitionClassTest.test_instance_field_getter`):

```
class C {
  int foo = 0;
  int get foo => 0;
}
```

The resolved AST contains an entry for the first instance of `foo` with a `nameOffset` of 16 (as expected), but its `.staticElement.getter` returns the explicit getter from the next line:

```
> field.name.offset
16
> (field.name.staticElement as PropertyInducingElement).getter!.nonSynthetic.nameOffset
35 // expected to also be 16?

> ((class.members[0] as FieldDeclaration).fields.variables[0].name.staticElement as PropertyInducingElement).getter
      == (class.members[1] as MethodDeclaration).name.staticElement
true
```

I don't know if this is intended or a bug, but while trying to report a `contextMessage` against duplicate definitions, this means we end up producing the same location in the `contextMessage` as the original diagnostic. It may be possible to keep hold of the original `identifier` to correct this, but that might not be necessary if this is not intended behaviour.

(cc @scheglov)

**Comment by bwilkerson** [comment]
Interestingly, navigation appears to be working correctly. Given the following expanded version of the code above

```dart
class C {
  int foo = 0;
  set foo(int x);
  int get foo => 0;

  void m() {
    foo;
    foo = 2;
  }
}
```

The two references to `foo` in `m` both navigate to the correct location. Given that navigation is based on the element model, that implies (though I haven't looked at the implementation) that suggests that the correct offset is still somewhere in the element model and that navigation support is accessing the correct offset.

If that's true, then there might be a work-around for this bug.

**Comment by DanTup** [comment]
It looks like computing navigation uses `writeOrReadElement` where this duplicate identifier uses `staticElement.getter`. Eg.:

```dart
      print(identifier.name);
      print(identifier.writeOrReadElement?.nameOffset);
      print((identifier.staticElement as PropertyInducingElement).getter?.nameOffset);
```

Outputs:

```
foo
16
35
```

For the original sample above. The use of [`.getter` in the duplicate check is from here](https://github.com/dart-lang/sdk/blob/a8c4ccc78347516881d32a62abc9fc91047efc93/pkg/analyzer/lib/src/error/duplicate_definition_verifier.dart#L488-489).

I don't understand these fields well enough to know whether there's an obvious workaround there though.

**Comment by scheglov** [comment]
The reason this happens is that we create `PropertyAccessorElement`s unconditionally, when we see either a `FieldDeclaration` or a `MethodDeclaration`. For a getter/setter `MethodDeclaration` we might create a synthetic `FieldElement` if it does not exist yet, which we check by the fact that there is an element for `Reference` like `@class::C::@field::foo`. In this case there is a `FieldElement`, so we don't create a new one. Then we unconditionally update `FieldElement` and set either `getter` or `setter` property.

This is why this code
```dart
  solo_test_XXX() async {
    await resolveTestCode(r'''
class C {
  int foo = 0;
  int get foo => 0;
}
''');

    final field = findNode.fieldDeclaration('foo = 0;');
    final variable = field.fields.variables[0];
    final varElement = variable.declaredElement as FieldElement;
    print(varElement.nameOffset);
    print(varElement.getter!.nameOffset);
  }
```
...prints
```
16
35
```

**Comment by DanTup** [comment]
> In this case there is a `FieldElement`, so we don't create a new one. Then we unconditionally update `FieldElement` and set either `getter` or `setter` property.

Is this the intended/desired functionality, or just a what's happening today?  (I'm wondering whether I should find a different way to handle the context messages, or whether the above should change - for example not reusing the same element as described above)?

**Comment by scheglov** [comment]
It was just happening today.
https://dart-review.googlesource.com/c/sdk/+/253362 should make it better.

**Comment by DanTup** [comment]
Great! That change fixes all my failing cases, thanks! :-)

**Comment by DanTup** [comment]
Fixed by 5e45794796051972c374956228ccd3d2d93a3cc9.
